x/valset-pref/validator_set.go

```diff
@@ -119,72 +128,125 @@ func (k Keeper) DelegateToValidatorSet(ctx sdk.Context, delegatorAddr string, co
+			UndelegateAmt: amountToUnDelegate,
+			VRatio:        valShares,
+			DelegatedAmt:  delegation.Shares.TruncateInt(),
+			Weight:        val.Weight,
+		})
+	}
+
+	// Step 3
+	sort.Slice(valSetRatio, func(i, j int) bool {
+		return valSetRatio[i].VRatio.GT(valSetRatio[j].VRatio)
+	})
+
+	// Step 4
+	totalUnDelAmt, amountToUnDelegate := sdk.NewInt(0), sdk.NewInt(0)
+	if valSetRatio[0].VRatio.LTE(sdk.OneDec()) {
+		fmt.Println("All validators can undelegate normally, falling to happy path exit early")
+		for index, val := range valSetRatio {
+			_, validator, err := k.getValAddrAndVal(ctx, val.ValAddr.String())
+			if err != nil {
+				return err
+			}
+
+			// in the last valset iteration we dont calculate it from shares using decimals and trucation,
+			// we use whats remaining to get more accurate value
+			if len(existingSet.Preferences)-1 == index {
+				amountToUnDelegate = coin.Amount.Sub(totalUnDelAmt).ToDec().TruncateInt()
+			} else {
+				// Calculate the amount to undelegate based on the existing weights
+				amountToUnDelegate = val.UndelegateAmt
+				totalUnDelAmt = totalUnDelAmt.Add(amountToUnDelegate)
+			}
+
+			sharesAmt, err := validator.SharesFromTokens(amountToUnDelegate)
+			if err != nil {
+				return err
+			}
+
+			_, err = k.stakingKeeper.Undelegate(ctx, delegator, val.ValAddr, sharesAmt) // this has to be shares amount
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}
+
+	// Step 5
+	// `targetRatio`: This is a threshold value that is used to decide how to unbond tokens from validators.
+	// It starts as 1 and is recalculated each time a validator is fully unbonded and removed from the unbonding process.
+	// By reducing the target ratio using the ratio of the removed validator, we adjust the proportions we are aiming for with the remaining validators.
+	targetRatio := sdk.OneDec()
+	amountRemaining := coin.Amount
+
+	fmt.Println("VRATIO PRE: ", valSetRatio)
+	// Step 6
+	for len(valSetRatio) > 0 && valSetRatio[0].VRatio.GT(targetRatio) {
+		fmt.Printf("Undelegating fully from validator %s %s\n", valSetRatio[0].ValAddr.String(), valSetRatio[0].DelegatedAmt)
+		_, err = k.stakingKeeper.Undelegate(ctx, delegator, valSetRatio[0].ValAddr, valSetRatio[0].DelegatedAmt.ToDec()) // this has to be shares amount
 		if err != nil {
 			return err
 		}
+		amountRemaining = amountRemaining.Sub(valSetRatio[0].DelegatedAmt)
+		targetRatio = targetRatio.Mul(sdk.OneDec().Sub(valSetRatio[0].Weight))
+		valSetRatio = valSetRatio[1:]
 	}
-	return nil
-}
 
-// CheckUndelegateTotalAmount checks if the tokenAmount equals the total amount calculated from valset weights.
-func (k Keeper) CheckUndelegateTotalAmount(tokenAmt sdk.Dec, existingSet []types.ValidatorPreference) error {
-	totalAmountFromWeights := sdk.NewDec(0)
-	for _, val := range existingSet {
-		totalAmountFromWeights = totalAmountFromWeights.Add(val.Weight.Mul(tokenAmt))
-	}
+	// Step 7
+	fmt.Println("Undelegating the remaining amount.")
```

Comment:
note: i'm a little worried about this part of the algorithm. would appreciate a thorough review here